popde/adaptive_kde.py

```diff
@@ -0,0 +1,191 @@
+import numpy as np
+from density_estimate import VariableBwKDEPy 
+from KDEpy.TreeKDE import TreeKDE
+
+class AdaptiveBwKDE(VariableBwKDEPy):
+    """
+    General adaptive Kernel Density Estimation (KDE) using KDEpy
+    with variable bandwidth per point
+
+    Example:
+    # Example of 3D KDE and plot for verification
+    mean1, sigma1 = 0.0, 1.0
+    mean2, sigma2 = 0.0, 1.0
+    mean3, sigma3 = 0.0, 1.0
+    n_samples = 1000
+    rndgen = np.random.RandomState(seed=1)
+    sample1 = rndgen.normal(mean1, sigma1, size=n_samples)
+    sample2 = rndgen.normal(mean2, sigma2, size=n_samples)
+    sample3 = rndgen.normal(mean3, sigma3, size=n_samples)
+    sample = np.column_stack((sample1, sample2, sample3)) # shape is (n_points, n_features)
+    # Create and fit the adaptive KDE note backend must be set correctly
+    kde = AdaptiveBwKDE(sample, backend='awKDEpy', dim_names=['x', 'y', 'z'], alpha=0.5, input_transf=None)
+    print("kde=", kde)
+    
+    # Generate grid for plotting
+    xgrid = np.linspace(sample1.min(), sample1.max(), 100)
+    ygrid = np.linspace(sample2.min(), sample2.max(), 100)
+    zgrid = np.linspace(sample3.min(), sample3.max(), 100)
+    XX, YY, ZZ = np.meshgrid(xgrid, ygrid, zgrid)
+    eval_pts = np.column_stack((XX.flatten(), YY.flatten(), ZZ.flatten()))
+    kde.fit()
+
+    # Evaluate the KDE at the grid points
+    density_values = kde.evaluate(eval_pts)
+    density_values = density_values.reshape(XX.shape)
+    print(density_values)
+    """
+    def __init__(self, data, backend='KDEpy', bandwidth=1., dim_names=None, alpha=0.5,  input_transf=None, stdize=False, rescale=None):
+        # Additional parameter for the Wang & Wang formula
+        self.alpha = alpha
+        super().__init__(data, input_transf, stdize, rescale,
+                         backend, bandwidth, dim_names)  
+
+    def _calc_local_bandwidth(self, kde_values):
+        """
+        Calculate local bandwidth using the provided function.
+
+        Parameters:
+        -----------
+        kde_values : array-like
+            The KDE values at the data point positions.
+
+        Returns:
+        --------
+        inv_loc_bw : array-like
+            Inverse of the local bandwidth calculated using the provided function.
```

Comment:
no, inv_loc_bw is not the inverse of the local bw, it is the inverse factor we apply to the global bw.  Also 'calculated using the provided function' does not give the reader any information.  If you want to explain _which_ function you could give a reference to an equation in the Wang paper

Reply:
changes done